pandera/schema_components.py

```diff
@@ -585,7 +585,19 @@ def coerce_dtype(self, obj: pd.MultiIndex) -> pd.MultiIndex:
             multiindex_cls = ps.MultiIndex
         return multiindex_cls.from_arrays(
             [
-                v.to_numpy()
+                # Context 1 : v.dtype may be different than 'object'.
```

Comment:
I think perhaps we could do an early return in the the pyspark special case in line 581 so that we don't need to special case it below on line 599. What do you think?

Reply:
@davidandreoletti friendly ping ^^

Reply:
@cosmicBboy At L583-L585, pyspark's MultiIndex implementation is assigned as class to use (not sure why the hack exist /  I did not investigate)
After that line, the goal is to fill **any** MultiIndex implementation provided with ideally "better typed array values" **for each** index column.

If I your request was:

```
if pyspark:
    return [ v.to_numpy() for k,v in sorted(...)]
else:
    return [ v.array for k,v in sorted(...) ]
```

Then, IMHO, I would keep the code as is because the L599 check works on the **input value** `v` and not the particular implementation of MultiIndex used `multiindex_cls`.

If tomorrow, another platform needs to call `v.some_other_array` then it needs to happen inside L598-L600.

Reply:
ah, I see, missed that. in that case would you mind using the stricter check `type(v).__module__.startswith("pyspark.pandas")` so that this is constrained to pyspark pandas?

In the next few release we'll be adding df-framework-specific backends so that this code isn't as messy 😅